File: 002.Add Two Numbers/Jimmy Xiang/002.go
package main

import (
	"container/list"
	"fmt"
)

func addTwoNumbers(a *list.List, b *list.List) (c *list.List) {

	d := list.New()

	for e1, e2 := a.Front(), b.Front(); e1 != nil && e2 != nil; e1, e2 = e1.Next(), e2.Next() {

		v1 := e1.Value.(int) //e1.Value 为 interface{}
		v2 := e2.Value.(int)

		value := (v1 + v2) % 10

		d.PushBack(value)

		if e1.Next() == nil {

			for e2.Next() != nil {

				d.PushBack(e2.Next().Value.(int))
				e2 = e2.Next()
			}
		}

		if e2.Next() == nil {

			for e1.Next() != nil {

				d.PushBack(e1.Next().Value.(int))
				e1 = e1.Next()
			}

		}

	}

	return d
}

func main() {

	a := list.New()
	b := list.New()
	c := list.New()

	a.PushBack(1)
	a.PushBack(2)
	a.PushBack(9)
	a.PushBack(4)

	b.PushBack(3)
	b.PushBack(7)

	c = addTwoNumbers(a, b)

	for e := c.Front(); e != nil; e = e.Next() {

		fmt.Println(e.Value)
	}
}

File: 002.Add Two Numbers/Jimmy Xiang/002.py
class ListNode:
    def __init__(self, v):
        self.val = v
        self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        head = ListNode(0)
        current, carry = head, 0

        while l1 or l2:
            val = carry
            if l1:
                val += l1.val
                l1 = l1.next
            if l2:
                val += l2.val
                l2 = l2.next

            carray,val = val / 10, val % 10
            current.next = ListNode(val)
            current = current.next

            if carray == 1:
                current.next = ListNode(1)

        return head.next

def main():
    a,a.next,a.next.next = ListNode(2),ListNode(4), ListNode(3)
    b, b.next, b.next.next = ListNode(5), ListNode(6), ListNode(4)
    result = Solution().addTwoNumbers(a, b)
    print "{0} -> {1} -> {2}".format(result.val, result.next.val, result.next.next.val)


if __name__ == "__main__":
    main()
